6.2-c : update libraries with optim config

guilpier-code opened the pull request.

Commit:
6.2c - updates libs with config : variables

Commit:
6.2c - updates libs with config : remove things now useless

Commit:
6.2c - updates libs with config : remove things now useless

Commit:
6.2c - updates libs with config : remove things now useless

Commit:
6.2c - update libs with config : objectives

flomnes commented on src/io/inputs/yml-system/converter.cpp:
```diff
@@ -101,7 +101,7 @@ static Component createComponent(const YmlSystem::Component& c,
     const auto [libraryId, modelId] = splitLibraryModelString(c.model);
 
-    const Model& model = getModel(libraries, libraryId, modelId);
+    Model& model = getModel(libraries, libraryId, modelId);
```
Note a non-const ref can be converted to `const`. A `const` ref can't be converted to non-const, except using `const_cast`.

You probably don't need to remove the `const` qualifier here.

guilpier-code replied:
Adding back a **const** causes a compilation error a bit further when calling  `component_builder.withModel(&model)`

guilpier-code replied:
There would be a solution to all of this **const** problematic.
We could first load libraries and option config under the form of a yml data structures, and then build business data structure with these.

a-zakir replied:
Have you examined the case of injecting optimConfig before libraries are created?

a-zakir replied:
mostly to avoid de-consting Model

guilpier-code replied:
Yes, but this way does not allow checking that all variables / objectives in the **optim-config.yml** exist in one of the libraries yml files

a-zakir replied:
after you mapped libraries with optimConfig , you'd need a simple loop on optimConfig and raise exception if there is orphan optimConfig  .vars or objectives

guilpier-code replied:
Indeed, we could fix it that way too.
I suggest we open a ticket on that matter and treat it at some point, as soon as possible.

guilpier-code replied:
Here is [the ticket](https://gopro-tickets.rte-france.com/browse/ANT-4051)

flomnes commented on src/solver/optim-model-filler/ComponentFiller.cpp:
```diff
@@ -424,7 +424,7 @@ void ComponentFiller::addObjectivesToMaster(const LinearProblemApi::FillContext&
 void ComponentFiller::addObjectives(const LinearProblemApi::FillContext& ctx)
 {
-    const auto* model = component_.getModel();
+    auto* model = component_.getModel();
```
The `const` qualifier can probably be kept

guilpier-code replied:
If we make this model **const**, then we cannot call **Objectives()** on it (turned to non-const function in this PR).
This causes a compilation error.

flomnes commented on src/io/inputs/yml-optim-config/converter.cpp:
```diff
@@ -50,69 +49,95 @@ Modeler::Config::Location convertLocation(const std::string& locationStr)
 {
-    std::vector<Modeler::Config::Objective> objectives;
-    for (const auto& obj: ymlModel.objectives)
+    auto filter = [&obj_id](const SystemModel::Objective& o) { return o.Id() == obj_id; };
```
```suggestion
    auto filter = [&obj_id](const SystemModel::Objective& objective) { return objective.Id() == obj_id; };
```

guilpier-code replied:
done

flomnes commented on src/io/inputs/yml-optim-config/converter.cpp:
```diff
@@ -50,69 +49,95 @@ Modeler::Config::Location convertLocation(const std::string& locationStr)
 
-std::vector<Modeler::Config::Objective> convertObjectives(const Model& ymlModel)
+
+SystemModel::Objective& findSystemObjective(std::string obj_id, SystemModel::Model& sysModel)
```
```suggestion
SystemModel::Objective& findSystemObjective(const std::string& obj_id, SystemModel::Model& sysModel)
```

guilpier-code replied:
done

flomnes commented on src/io/inputs/yml-optim-config/converter.cpp:
```diff
@@ -50,69 +49,95 @@ Modeler::Config::Location convertLocation(const std::string& locationStr)
 
-std::vector<Modeler::Config::Variable> convertVariables(const Model& ymlModel)
+SystemModel::Variable& findSystemVariable(std::string variable_id, SystemModel::Model& sysModel)
 {
```
```suggestion
SystemModel::Variable& findSystemVariable(const std::string& variable_id, SystemModel::Model& sysModel)
{
```

guilpier-code replied:
Done

flomnes commented on src/io/inputs/yml-optim-config/include/antares/io/inputs/yml-optim-config/converter.h:
```diff
@@ -23,14 +23,16 @@
 public:
-    static Modeler::Config::OptimConfig convert(const OptimConfig& ymlOptimConfig);
+    static void updateLibrairies(const OptimConfig& ymlOptimConfig,
+                                 std::vector<ModelerStudy::SystemModel::Library>& libraries);
```
A class with a single `static` method looks strange, is the `class` really needed ?

guilpier-code replied:
Indeed. Turned into a free function.

flomnes commented on src/io/inputs/yml-optim-config/converter.cpp:
```diff
@@ -50,69 +49,95 @@ Modeler::Config::Location convertLocation(const std::string& locationStr)
+    if (search == lib->Models().end())
     {
-        throw Error::RuntimeError("Error converting objective '" + ymlObj.id + "': " + e.what());
+        throw std::runtime_error("No library found with this name: " + modelId);
```
Typo
```suggestion
        throw std::runtime_error("No model found with this name: " + modelId);
```

guilpier-code replied:
Done

flomnes commented on src/io/inputs/yml-optim-config/converter.cpp:
```diff
@@ -50,69 +49,95 @@ Modeler::Config::Location convertLocation(const std::string& locationStr)
     {
-        objectives.push_back(convertObjective(obj));
-        std::vector<Modeler::Config::Objective> objectives;
+        throw std::runtime_error("No objetctive found with this name: " + obj_id);
```
Typo
```suggestion
        throw std::runtime_error("No objective found with this name: " + obj_id);
```

guilpier-code replied:
Done

flomnes reviewed (commented):
LGTM

A few typos

We should add unit tests for this

Commit:
6.2c - update libs with config : move optimConfig.h et consequences

Commit:
6.2c - update libs with config : fix

Commit:
6.2c - update libs with config : changes due to review

Commit:
6.2c - update libs with config : turn a member function into a free function

Commit:
6.2c - update libs with config : fixes due to review

guilpier-code committed:
6.2-c : fixes on optim-config.yml (#3202)

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>

Commit:
6.2c - update libs with config : cleaning

Commit:
Merge remote-tracking branch 'remotes/origin/feature/6.2c--step-2' into feature/6.2c--step-3

guilpier-code committed:
6.2-c : Fixes on objective contributions (#3203)

Asked by @tbittar

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>

Commit:
6.2-c : Fixes on objective contributions

flomnes committed:
Add test coverage for variable & objective decomposition (#3204)

- Add unit test for variables
- Allow absence of attribute `variables` and `objective-contributions`
when parsing optim-config.yml

```yaml
models:
  - id: library.model
    model-decomposition:
      variables:
        - id: y
          location: master-and-subproblems
```
